### PR TITLE
Nginx plugin

### PR DESCRIPTION
**Pull request by dev0z**
- Nginx connector to use waflz lib.
- Supports scopes sans rl and bot challenge
- nginx plugin still in poc mode. Need to add more callbacks if refactoring is approved
- Refactored rqst_ctx callbacks. From references -> pointers to work with C libs
- fixed a bug in which parsers were always turned off in rqst_ctx creation in scopes.cc

### CODE REVIEW

File: include/waflz/def.h [hunk 1]
```diff
@@ -72,20 +79,28 @@
...
-typedef int32_t (*get_rqst_data_size_cb_t)(uint32_t &, void *);
-typedef int32_t (*get_rqst_data_cb_t)(const char **, uint32_t &, void *);
-typedef int32_t (*get_rqst_data_w_key_cb_t)(const char **, uint32_t &, void *, const char *, uint32_t);
-typedef int32_t (*get_rqst_kv_w_idx_cb_t)(const char **, uint32_t &, const char **, uint32_t &, void *, uint32_t);
-typedef int32_t (*get_rqst_body_data_cb_t)(char *, uint32_t &, bool &, void *, uint32_t);
+typedef int32_t (*get_rqst_data_size_cb_t)(uint32_t *a, void *);
+typedef int32_t (*get_rqst_data_cb_t)(const char **, uint32_t *, void *);
+typedef int32_t (*get_rqst_data_w_key_cb_t)(const char **, uint32_t *, void *, const char *, uint32_t);
+typedef int32_t (*get_rqst_kv_w_idx_cb_t)(const char **, uint32_t *, const char **, uint32_t *, void *, uint32_t);
+typedef int32_t (*get_rqst_body_data_cb_t)(char *, uint32_t *, bool , void *, uint32_t *);
```
Comment (RevathiSaba) [inline include/waflz/def.h:101]: fourth parameter is pointer intentionally or typo?

File: plugins/nginx/src/ngx_http_waflz_module.c [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+#include "ngx_http_waflz_module.h"
+
+static ngx_int_t ngx_http_waflz_init(ngx_conf_t *cf);
+//: ----------------------------------------------------------------------------
+//: routines to allocate and init main conf
+//: ----------------------------------------------------------------------------
+static void * ngx_http_waflz_create_main_conf(ngx_conf_t *cf);
```
Comment (RevathiSaba) [inline plugins/nginx/src/ngx_http_waflz_module.c]: i think pointer * can be either near void/type or near function.. i always do near type but reed does it near function..google c++ std is near type i think. so whichever you prefer.. we can start making it consistent in new files

File: src/core/scopes.cc [hunk 3]
```diff
@@ -1086,6 +1087,32 @@ int32_t scopes::load_parts(waflz_pb::scope& a_scope,
...
+                                       void *a_ctx,
+                                       const rqst_ctx_callbacks *a_callbacks,
+                                       rqst_ctx **ao_rqst_ctx)
+{
+        waflz_pb::event *l_audit_event = NULL;
+        waflz_pb::event *l_prod_event = NULL;
+        const waflz_pb::enforcement *l_enf = NULL;
+        int32_t l_s;
+        l_s = process(&l_enf, &l_audit_event, &l_prod_event, a_ctx, PART_MK_ALL, a_callbacks, ao_rqst_ctx);
+        // TODO: dev handle both audit and prod
```
Comment (RevathiSaba) [inline src/core/scopes.cc:1104]: is this TODO for later?

Comment (RevathiSaba) [reply src/core/scopes.cc:1104]: i mean for next PR?

File: plugins/nginx/src/ngx_http_waflz_module.c (hunk 2 above)
Comment (dev0z) [reply plugins/nginx/src/ngx_http_waflz_module.c]: thx, changed it to nginx style guide. which is near func/var

File: src/core/scopes.cc (hunk 3 above)
Comment (dev0z) [reply src/core/scopes.cc:1104]: Yeah, lot of TODOs, for now just in a POC mode to send an alert. Will follow up with more changes to this func

File: include/waflz/def.h (hunk 1 above)
Comment (dev0z) [reply include/waflz/def.h:101]: intentionally made it to a pointer instead of pass by value. Could have left it as it is.

**Review by RevathiSaba — APPROVED** [review]
lgtm

**Pull request merged**